File: java/src/main/java/br/datastructures/queue/LinkedQueue.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package br.datastructures.queue;

import br.datastructures.exceptions.NoSuchItemException;

public class LinkedQueue<I> extends AbstractQueue<I> {
    class Node {
        I item;
        Node next;
        
        Node(I item, Node next) {
            this.item = item;
            this.next = next;
        }
        
        Node(I item) {
            this(item, null);
        }
    }
    
    private Node first, last;

    public LinkedQueue() {
        first = null;
        last = null;
    }
    
    @Override
    public void enqueue(I item) {
        Node newNode = new Node(item);
        
        if (empty()) {
            first = newNode;
            last = newNode;
        }
        else {
            last.next = newNode;
            last = newNode;
        } 
        
        quantity++;
    }
    
    @Override
    public I getFirst() throws NoSuchItemException {
        if(empty())
            throw new NoSuchItemException();
        
        return first.item;
    }
    
    @Override
    public I dequeue() throws NoSuchItemException {
        I item = getFirst();
        first = first.next;
        
        if(first == null) 
            last = null;
        
        quantity--;

        return item;
    }
}

File: c/queue/LinkedQueue.c
#include <stdio.h>
#include <stdlib.h>

#define NENHUM_ITEM -1

typedef int Item;

typedef struct No {
    Item item;
    struct No* next;
} No;

typedef struct {
    No* first;
    No* last;
    int quantity;
} LinkedQueue;

LinkedQueue* newLinkedQueue() {
    LinkedQueue* queue = (LinkedQueue*) malloc(sizeof(LinkedQueue));

    if(queue == NULL)
        return NULL;

    queue->first = NULL;
    queue->last = NULL;
    queue->quantity = 0;

    return queue;
}

int size(LinkedQueue* queue) {
    return queue->quantity;
}

int empty(LinkedQueue* queue) {
    return size(queue) == 0;
}

void enqueue(LinkedQueue* queue, Item item) {
    No* no = (No*) malloc(sizeof(No));
    if(no == NULL)
        return;
        
    no->item = item;
    no->next = NULL;

    if(empty(queue)) 
        queue->first = no;
    else 
        queue->last->next = no;

    queue->last = no;
    queue->quantity++;
}

Item getFirst(LinkedQueue* queue) {
    if(empty(queue))
        return NENHUM_ITEM;

    return queue->first->item;
}

Item dequeue(LinkedQueue* queue) {
    if(empty(queue))
        return NENHUM_ITEM;
    
    Item item = getFirst(queue);

    No* first = queue->first;
    queue->first = queue->first->next;
    queue->quantity--;

    if(empty(queue))
        queue->last = NULL;

    free(first);

    return item;
}

void viewQueue(LinkedQueue* queue) {
    if(empty(queue)) {
        printf("Fila vazia!!\n");
        return;
    }

    printf("Itens da fila:");
    while(!empty(queue))
        printf(" %d", dequeue(queue));
    
    printf("\n");
}

void freeQueue(LinkedQueue* queue) {
    while(!empty(queue))
        dequeue(queue);
    
    free(queue);
}

int main() {
    LinkedQueue* queue = newLinkedQueue();

    enqueue(queue, 10);
    printf("Primeiro item: %d\n", getFirst(queue));

    enqueue(queue, 11);
    printf("Primeiro item: %d\n", getFirst(queue));

    enqueue(queue, 12);
    printf("Primeiro item: %d\n", getFirst(queue));

    printf("%d removido\n", dequeue(queue));

    viewQueue(queue);

    freeQueue(queue);

    return 0;
}